Bacteria v-1.0

besperspektivnyak opened the pull request:
Первая (черновая) реализация классов Bacteria и GodSimulator. Немного иначе, чем на схеме, работают функции из-за возможного конфликта при взаимном подключении пакетов друг в друга.

besperspektivnyak committed:
add first realization

first (draft) realization of Bacteria and GodSimulator classes

besperspektivnyak committed:
Delete README.md

MaEgV commented on GodSimulator.py:
```diff
@@ -0,0 +1,23 @@
+
+    @staticmethod
+    def have_to_reproduct(p: BacteriaParameters):
+        if p.p_for_separate < random.random:
```
return p.p_for_separate > random.random

MaEgV commented on Bacteria.py:
```diff
@@ -0,0 +1,22 @@
+        self.parameters = p
+
+    def iteration(self):
+        children = list()
```
Нет проверки на смерть бактерии

MaEgV commented on BacteriaParameters.py:
```diff
@@ -0,0 +1,8 @@
+class BacteriaParameters:
+
+    def __init__(self, id_: int, max__life_time_: int, p_for_death_: float, p_for_reproduction_: float):
+        self.id = id_
```
id можно убрать. Ошибка проектирования

besperspektivnyak committed:
fix errors

fix minor errors in first version

MaEgV commented on Bacteria.py:
```diff
@@ -13,10 +13,14 @@ def __init__(self, p: BacteriaParameters):
-        child_parameters = GodSimulator.get_child_parameters(self.parameters)
-        while child_parameters:
-            children.append(create_child(child_parameters))
+        if self.is_alive:
```
Пропущена проверка GodSimulator.have_to_die

MaEgV commented on GodSimulator.py:
```diff
@@ -13,7 +13,7 @@ def have_to_die(p: BacteriaParameters):
     @staticmethod
     def have_to_reproduct(p: BacteriaParameters):
-        if p.p_for_separate < random.random:
+        if p.p_for_separate > random.random:
```
Выражение можно сократить до return p.p_for_separate > random.random()

MaEgV replied:
p_for_reproduction*

MaEgV reviewed (commented):
This will be a great basic version after the fixes

MaEgV committed:
Typos fix

Resolving conflicts with other classes. Adding brackets

MaEgV commented on Bacteria.py:
```diff
@@ -13,10 +13,14 @@ def __init__(self, p: BacteriaParameters):
+            children = list()
             child_parameters = GodSimulator.get_child_parameters(self.parameters)
-        return children
+            while child_parameters:
```
while GodSimulator.have_to_reproduct(p):
child_parameters = GodSimulator.get_child_parameters(self.parameters)
children.append(create_child(child_parameters))

besperspektivnyak committed:
fix logical errors

besperspektivnyak committed:
remove redundant strings

besperspektivnyak committed:
добавлены новые и обновленные файлы

изменена структура бактерии, усложнены принципы, по которым наследуются свойства

MaEgV commented on Bacteria.py:
```diff
@@ -0,0 +1,22 @@
+from GodSimulator import GodSimulator
+
+
+def create_child(p: BacteriaParameters):
```
Прошу сделать из этой функции следующую: create_bacteria(**kwargs) -> Bacteria

MaEgV commented on Bacteria.py:
```diff
@@ -0,0 +1,22 @@
+        children = list()
+        child_parameters = GodSimulator.get_child_parameters(self.parameters)
+        while child_parameters:
+            children.append(create_child(child_parameters))
```
Поменять на Bacteria(child_parameters)

MaEgV commented on GodSimulator.py:
```diff
@@ -1,17 +1,35 @@
 import BacteriaParameters
 
 
+def probability_parameter(parent_parameter: float):
+    probability_variation = uniform(-0.1, 0.1)
+    parameter = parent_parameter + probability_variation
+    if parameter > 1:
+        parameter = 1
+    return parameter
+
+
+def choose_child_parameters(parent_parameters: BacteriaParameters):
+    child_parameters = copy.deepcopy(parent_parameters)
+    child_parameters.genome.p_for_death = probability_parameter(parent_parameters.genome.p_for_death)
+    child_parameters.genome.p_for_reproduction = probability_parameter(parent_parameters.genome.p_for_reproduction)
+    child_parameters.genome.max_n = randint(0, 15)
+    return child_parameters
```
Внести функции в класс как приватные и добавить константы, как public параметры со значениями по умолчанию
Или как приватные с set-ерами

besperspektivnyak committed:
правка второй версии

добавлены возвращаемые значения, изменены некоторые имена методов на более корректные, константы вынесены как отдельные поля класса

MaEgV committed:
Фикс незначительной логической ошибки

Особи с нулём оставшихся итераций быть не может

MaEgV committed:
Опечатки

besperspektivnyak committed:
добавлены комментарии

добавлены комментарии к классам и методам, изменен метода create_bacetria

besperspektivnyak committed:
удален GodSimulator.py

GodSimulator.py был заменен на EvolutionSimulator.py

MaEgV committed:
Typos fix
